src/test/fuzz/p2p_transport_serialization.cpp

```diff
@@ -63,22 +65,260 @@ FUZZ_TARGET(p2p_transport_serialization, .init = initialize_p2p_transport_serial
+                // The m_type must match what is expected.
+                assert(received.m_type == expected[side].front().m_type);
+                // The data must match what is expected.
+                assert(MakeByteSpan(received.m_recv) == MakeByteSpan(expected[side].front().data));
+                expected[side].pop_front();
+                progress = true;
+            }
+            // Progress must be made (by processing incoming bytes and/or returning complete
+            // messages) until all received bytes are processed.
+            assert(progress);
+        }
+        // Remove the processed bytes from the in_flight buffer.
+        in_flight[side].erase(in_flight[side].begin(), in_flight[side].begin() + to_recv_len);
+        // Return whether anything was received.
+        return to_recv_len > 0;
+    };
+
+    // Main loop, interleaving new messages, sends, and receives.
+    unsigned iter = 0;
+    while (iter < 1000 && provider.remaining_bytes()) {
+        uint8_t code = provider.ConsumeIntegral<uint8_t>();
+        // Lowest bit identifies the sender side of the operation.
+        int side = code & 1;
+        switch ((side >> 1) % 3) {
+        // The other bits (mod 3) identify the operation:
+        case 0:
+            // (Try to) give the next message to the transport.
+            new_msg_fn(side);
+            break;
+        case 1:
+            // (Try to) send some bytes from the transport to the network.
+            send_fn(side);
+            break;
+        case 2:
+            // (Try to) receive bytes from the network, converting to messages.
+            recv_fn(side);
+            break;
+        }
+        ++iter;
+    }
+
+    // When we're done, perform sends and receives of existing messages to flush anything already
+    // in flight.
+    while (true) {
+        bool any = false;
+        if (send_fn(0, true)) any = true;
+        if (send_fn(1, true)) any = true;
+        if (recv_fn(0, true)) any = true;
+        if (recv_fn(1, true)) any = true;
+        if (!any) break;
     }
+
+    // Make sure nothing is left in flight.
+    assert(in_flight[0].empty());
+    assert(in_flight[1].empty());
+
+    // Make sure all expected messages were received.
+    assert(expected[0].empty());
+    assert(expected[1].empty());
+}
+
+std::unique_ptr<Transport> MakeV1Transport(NodeId nodeid) noexcept
+{
+    return std::make_unique<V1Transport>(nodeid, SER_NETWORK, INIT_PROTO_VERSION);
+}
+
+} // namespace
+
+FUZZ_TARGET(p2p_transport_bidirectional, .init = initialize_p2p_transport_serialization)
+{
+    // Test with two V1 transports talking to each other.
+    FuzzedDataProvider provider{buffer.data(), buffer.size()};
+    XoRoShiRo128PlusPlus rng(provider.ConsumeIntegral<uint64_t>());
+    auto t1 = MakeV1Transport(NodeId{0});
+    auto t2 = MakeV1Transport(NodeId{1});
+    if (!t1 || !t2) return;
+    SimulationTest(*t1, *t2, rng, provider);
```

Comment:
Is there a reason to use pointers? The following seems to be simpler:

```cpp
    V1Transport t0{NodeId{0}, SER_NETWORK, INIT_PROTO_VERSION};
    V1Transport t1{NodeId{1}, SER_NETWORK, INIT_PROTO_VERSION};
    SimulationTest(t0, t1, rng, provider);
```
and remove `MakeV1Transport()`?

Reply:
It is ok as it is - there is _some_ reason for using pointers.